### ISSUE TITLE
first buildir run takes over a minute

### ISSUE THREAD

**Issue by tbrisker** [dominikh/go-tools#1171]
Hello,

We're running staticcheck from inside golangci-lint, and we're noticing that the first `buildir` run takes over a minute:

```
INFO [linters context/goanalysis] analyzers took 1m35.207052217s with top 10 stages: buildir: 1m14.632792599s, nilness: 7.243121049s, unconvert: 1.805017404s, gofmt: 1.491275756s, gosec: 1.403823546s, fact_purity: 958.863702ms, typedness: 909.852439ms, SA5012: 784.763919ms, misspell: 723.1936ms, SA1012: 275.686891ms 
```
Following runs finish very quickly as they do not rerun the analyzers (as the results are cached). 
However, since we're running the linters inside CI we start with a clean cache on every run.

This is golangci-lint v1.43.0 which runs version 0.2.0 of staticcheck, running on go 1.17.5.

Thanks in advance for any pointers on how we could make the `buildir` step run faster.

**Comment by dominikh** [comment]
> Thanks in advance for any pointers on how we could make the buildir step run faster.

This sounds like a pathological case, and I would need to take a look at the code you're analyzing to see what's causing it.

> This is golangci-lint v1.43.0 which runs version 0.2.0 of staticcheck, running on go 1.17.5.

Please test if the same issue occurs with `honnef.co/go/tools/cmd/staticcheck`. That will also allow you to use `-debug.measure-analyzers=output.txt` to produce per-analyzer per-package timing information, which should tell you which package(s) contribute significantly to the runtime, assuming the issue reproduces with cmd/staticcheck.

> However, since we're running the linters inside CI we start with a clean cache on every run.

Note that this isn't a given. Most CI environments can be configured to reuse caches between runs.

**Comment by tbrisker** [comment]
Thanks for the quick response @dominikh, much appreciated!
Unfortunately, this is an internal repo so I cannot share the code itself.
However, I did manage to get some statistics using staticcheck directly with the debug measurement flag running locally on my laptop and with a fresh cache. 
We seem to have a bit over 1100 packages being analyzed. 
Buildir in total takes about 45 seconds, with the average time of 39ms per package - although they are far from being evenly distributed. The top 20 packages account for 15 seconds, and the top 100 packages account for about 30 seconds.
The slowest package seems to be `github.com/aws/aws-sdk-go-v2/service/ec2` which takes 3 seconds for buildir to parse.

> Note that this isn't a given. Most CI environments can be configured to reuse caches between runs.
Thanks for the hint, I'll also look into the possibility of sharing the cache between builds.

**Comment by dominikh** [comment]
> Unfortunately, this is an internal repo so I cannot share the code itself.

I'd be happy to jump through legal hoops to look at it, too.

> although they are far from being evenly distributed

Running with GOMAXPROCS=1 helps reduce scheduling noise and should make for much better results.

**Comment by dominikh** [comment]
I think the ec2 package is a red herring. That package is _big_, containing 12940 functions in the IR representation. On my system, it takes 2s to build, which makes for 153 µs per function. In comparison, `runtime` has 1932 functions and takes 227ms, or 117 µs per function.

I've computed some basic statistics of running Staticcheck against `std`. 11.15s of 34.34s is spent in buildir, while the rest is spent in all other analyzes. I would expect to see a similar distribution on your code base.

To be frank, I don't trust the statistics emitted by golangci-lint. `analyzers took 1m35.207052217s with top 10 stages` is ambiguous. Does it mean 1m35s in total, or 1m35s spent in these 10 stages? Neither makes sense to me. Summing the 10 stages results in 1m30s, which is either inaccurate, or would mean only 5s left for all other analyzers, of which there are over 100.

Another comparison you could make: with empty caches, see how long it takes to run `staticcheck std` and `staticcheck your/codebase...`, and compare the number of packages (as determined by `go list std` and `go list your/codebase...`), and see if the ratios are roughly the same. If they are, then there's nothing unexpected happening.

**Comment by tbrisker** [comment]
> To be frank, I don't trust the statistics emitted by golangci-lint. 

I agree, the numbers don't quite add up. It might be total cpu time with some of the code being executed in parallel, but not sure. In any case, the clock time was a bit over a minute for the full run.

> I think the ec2 package is a red herring. That package is big, containing 12940 functions in the IR representation. On my system, it takes 2s to build, which makes for 153 µs per function. In comparison, runtime has 1932 functions and takes 227ms, or 117 µs per function. ...  If they are, then there's nothing unexpected happening.

That makes sense. This is similar to what I got, I just wasn't sure if that was the expected times or if there was room for some optimization. 
I suppose that we are seeing the long timing due to the number and size of the packages being analyzed in our project, and there isn't anything to be done on the staticcheck side about it.

By taking your advice on reusing the cache, I've been able to reduce the total linting run-time from ~1.5 minutes to ~10 seconds. 
Thank you very much for the effort to investigate and for the hint regarding reusing the cache!
